src/librbd/io/SimpleSchedulerObjectDispatch.cc

```diff
@@ -0,0 +1,359 @@
+    Context** on_finish) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  for (auto &it : m_delayed_requests) {
+    dispatch_delayed_requests(it.first, on_finish);
+  }
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    uint64_t object_no, Context** on_finish) {
+  ceph_assert(m_lock.is_locked());
+
+  dispatch_delayed_requests(object_no);
+
+  auto res = m_delayed_requests.insert({object_no, {object_no}});
+  ceph_assert(res.second);
+  auto it = res.first;
+
+  auto dispatch_seq = ++m_dispatch_seq;
+  it->second.dispatch_seq = dispatch_seq;
+  auto ctx = *on_finish;
+  *on_finish = new FunctionContext(
+    [this, object_no, dispatch_seq, ctx](int r) {
+      ctx->complete(r);
+
+      Mutex::Locker locker(m_lock);
+      auto it = m_delayed_requests.find(object_no);
+      if (it == m_delayed_requests.end() ||
+          it->second.dispatch_seq != dispatch_seq) {
+        ldout(m_image_ctx->cct, 20) << "already dispatched" << dendl;
+        return;
+      }
+      dispatch_delayed_requests(it->second);
+      m_delayed_requests.erase(it);
+    });
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    uint64_t object_no) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+
+  auto it = m_delayed_requests.find(object_no);
+  if (it == m_delayed_requests.end()) {
+    ldout(cct, 20) << "object_no=" << object_no << ": not found" << dendl;
+    return;
+  }
+
+  dispatch_delayed_requests(it->second);
+  m_delayed_requests.erase(it);
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    ObjectDelayedRequests &delayed_requests) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+
+  ldout(cct, 20) << "object_no=" << delayed_requests.object_no << ", "
+                 << delayed_requests.requests.size() << " requests" << dendl;
+
+  if (delayed_requests.requests.empty()) {
+    return;
+  }
+
+  auto requests = std::move(delayed_requests.requests);
+  auto ctx = new FunctionContext(
+    [this, requests](int r) {
+      for (auto on_dispatched : requests) {
+        on_dispatched->complete(r);
+      }
+    });
+
+  auto req = io::ObjectDispatchSpec::create_write(
+      m_image_ctx, io::OBJECT_DISPATCH_LAYER_SCHEDULER,
+      m_image_ctx->get_object_name(delayed_requests.object_no),
+      delayed_requests.object_no, delayed_requests.offset,
+      std::move(delayed_requests.data), delayed_requests.snapc,
+      delayed_requests.op_flags, 0, {}, ctx);
+
+  req->object_dispatch_flags = delayed_requests.object_dispatch_flags;
+  req->send();
+
+  ceph_assert(!m_dispatch_queue.empty());
+  if (m_dispatch_queue.front() == &delayed_requests) {
+    m_dispatch_queue.pop_front();
+    rescheduler_dispatch_delayed_requests();
+  } else {
+    m_dispatch_queue.remove(&delayed_requests);
```

Comment:
This seems like it could get expensive if you have a lot of in-flight IO to different objects. Perhaps you can use a boost::bimap to index by dispatch time **and** object number (or more generalized dispatch request sequence if this can be modified to support more than one queued write to an object).